command/agent/sink/file/file_sink.go

```diff
@@ -0,0 +1,112 @@
+package file
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/hashicorp/errwrap"
+	hclog "github.com/hashicorp/go-hclog"
+	uuid "github.com/hashicorp/go-uuid"
+	"github.com/hashicorp/vault/command/agent/sink"
+)
+
+// fileSink is a Sink implementation that writes a token to a file
+type fileSink struct {
+	path   string
+	logger hclog.Logger
+}
+
+// NewFileSink creates a new file sink with the given configuration
+func NewFileSink(conf *sink.SinkConfig) (sink.Sink, error) {
+	if conf.Logger == nil {
+		return nil, errors.New("nil logger provided")
+	}
+
+	conf.Logger.Info("creating file sink")
+
+	f := &fileSink{
+		logger: conf.Logger,
+	}
+
+	pathRaw, ok := conf.Config["path"]
+	if !ok {
+		return nil, errors.New("'path' not specified for file sink")
+	}
+	path, ok := pathRaw.(string)
+	if !ok {
+		return nil, errors.New("could not parse 'path' as string")
+	}
+
+	f.path = path
+
+	if err := f.WriteToken(""); err != nil {
+		return nil, errwrap.Wrapf("error during write check: {{err}}", err)
+	}
+
+	f.logger.Info("file sink configured", "path", f.path)
+
+	return f, nil
+}
+
+// WriteToken implements the Server interface and writes the token to a path on
+// disk. It writes into the path's directory into a temp file and does an
+// atomic rename to ensure consistency. If a blank token is passed in, it
+// performs a write check but does not write a blank value to the final
+// location.
+func (f *fileSink) WriteToken(token string) error {
+	f.logger.Trace("enter write_token", "path", f.path)
+	defer f.logger.Trace("exit write_token", "path", f.path)
+
+	u, err := uuid.GenerateUUID()
+	if err != nil {
+		return errwrap.Wrapf("error generating a uuid during write check: {{err}}", err)
+	}
+
+	targetDir := filepath.Dir(f.path)
+	fileName := filepath.Base(f.path)
+	tmpSuffix := strings.Split(u, "-")[0]
+
+	tmpFile, err := os.OpenFile(filepath.Join(targetDir, fmt.Sprintf("%s.tmp.%s", fileName, tmpSuffix)), os.O_WRONLY|os.O_CREATE, 0640)
```

Comment:
`ioutil.TempFile()` is another option

Reply:
That doesn't allow picking the mode, which eventually I'd like to add.